src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.c

```diff
@@ -0,0 +1,114 @@
+        (*env)->CallStaticObjectMethod(env, g_SignatureClass, g_SignatureGetInstance, algorithmName);
+    (*env)->DeleteLocalRef(env, algorithmName);
+    if (CheckJNIExceptions(env))
+        return NULL;
+    return signatureObject;
+}
+
+int32_t CryptoNative_EcDsaSign(const uint8_t* dgst, int32_t dgstlen, uint8_t* sig, int32_t* siglen, EC_KEY* key)
+{
+    if (!siglen)
+    {
+        return FAIL;
+    }
+
+    JNIEnv* env = GetJNIEnv();
+
+    jobject signatureObject = CryptoNative_GetEsDsaSignatureObject(env);
+    if (!signatureObject)
+    {
+        return FAIL;
+    }
+
+    jobject privateKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPrivateMethod);
+    (*env)->CallVoidMethod(env, signatureObject, g_SignatureInitSign, privateKey);
+    ReleaseLRef(env, privateKey);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray digestArray = (*env)->NewByteArray(env, dgstlen);
+    (*env)->SetByteArrayRegion(env, digestArray, 0, dgstlen, (const jbyte*)dgst);
+    (*env)->CallVoidMethod(env, signatureObject, g_SignatureUpdate, digestArray);
+    ReleaseLRef(env, digestArray);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray sigResult = (*env)->CallObjectMethod(env, signatureObject, g_SignatureSign);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+    jsize sigSize = (*env)->GetArrayLength(env, sigResult);
+    *siglen = sigSize;
+    (*env)->GetByteArrayRegion(env, sigResult, 0, sigSize, (jbyte*)sig);
+    ReleaseLRef(env, sigResult);
+
+    ReleaseLRef(env, signatureObject);
+
+    return SUCCESS;
+
+error:
+    ReleaseLRef(env, signatureObject);
+    return FAIL;
+}
+
+int32_t CryptoNative_EcDsaVerify(const uint8_t* dgst, int32_t dgstlen, const uint8_t* sig, int32_t siglen, EC_KEY* key)
+{
+    JNIEnv* env = GetJNIEnv();
+
+    jobject signatureObject = CryptoNative_GetEsDsaSignatureObject(env);
+    if (!signatureObject)
+    {
+        return FAIL;
+    }
+
+    jobject publicKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPublicMethod);
+    (*env)->CallVoidMethod(env, signatureObject, g_SignatureInitVerify, publicKey);
+    ReleaseLRef(env, publicKey);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray digestArray = (*env)->NewByteArray(env, dgstlen);
+    (*env)->SetByteArrayRegion(env, digestArray, 0, dgstlen, (const jbyte*)dgst);
+    (*env)->CallVoidMethod(env, signatureObject, g_SignatureUpdate, digestArray);
+    ReleaseLRef(env, digestArray);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray sigArray = (*env)->NewByteArray(env, siglen);
+    (*env)->SetByteArrayRegion(env, sigArray, 0, siglen, (const jbyte*)sig);
+    jboolean verified = (*env)->CallBooleanMethod(env, signatureObject, g_SignatureVerify, sigArray);
+    LOG_DEBUG("Ran verification.");
+    ReleaseLRef(env, sigArray);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    ReleaseLRef(env, signatureObject);
+
+    return verified ? SUCCESS : FAIL;
+
+error:
+    ReleaseLRef(env, signatureObject);
+    return -1;
```

Comment:
How is this different from `FAIL` above?

Reply:
Like RSA, some of these functions on OpenSSL return 1 for success, 0 for completion but failure (like a failed signature verification), and -1 for error.

I’ve adopted the same convention here for simplicity since the managed side still expects the same codes.